File: MtxVec.CPP/BenchmarkX.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "BenchmarkX.h"
#include "MtxVec.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TBenchmarkXForm *BenchmarkXForm;
//---------------------------------------------------------------------------
__fastcall TBenchmarkXForm::TBenchmarkXForm(TComponent* Owner)
  : TBasicForm2(Owner)
{
    fFirstActivation = true;
    fBenchmarkFramework = new BenchmarkFramework();
    fBenchmarkResults = new TBenchmarkResults();
    LoadFuncListBox();

    AnsiString str;

    RichEdit1->Lines->Clear();
    str = "These benchmarks compare vectorized and not vectorized ";
    str+= "versions of real and complex versions of basic math functions. ";
    str+= "The vectorized versions are labeled with MtxVec and the non-vectorized ";
    str+= "versions are labeled with Math387.";
    RichEdit1->Lines->Add(str);
    RichEdit1->Lines->Add("");
    str = "The non-vectorized real and complex versions ";
    str+= "are written in assembler for maximum performance. ";
    str+= "Despite this, the vectorized versions are substantially faster. ";
    str+= "Before running the tests double click on the iterations combobox. ";
    str+= "That will compute the number of iterations that have to be performed ";
    str+= "to get sufficiently long timings. The resolution of the timings is about 15ms.";
    RichEdit1->Lines->Add(str);
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::FormDestroy(TObject *Sender)
{
    delete fBenchmarkFramework;
    delete fBenchmarkResults;
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::ShowResults(TBenchmarkResults *results)
{
    UpdateChart();
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::LoadFuncListBox()
{
  int i;

  clbFuncs->Items->BeginUpdate();
  try {
    clbFuncs->Items->Clear();

    for (i = 0; i < fBenchmarkFramework->FuncCount; i++) {

      clbFuncs->Items->Add (fBenchmarkFramework->FuncName[i]);
      clbFuncs->Checked[i] = true;
    }
  }
  __finally {
    clbFuncs->Items->EndUpdate();
  }
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::SelectAllFuncs(bool select)
{
  int i;

  for (i = 0; i < clbFuncs->Items->Count; i++) {
    clbFuncs->Checked[i] = select;
  }
}

//---------------------------------------------------------------------------


void __fastcall TBenchmarkXForm::clbFuncsClick(TObject *Sender)
{
  int idx;

  idx = clbFuncs->ItemIndex;
  if (idx != -1) {

      clbFuncs->Checked[idx] = !(clbFuncs->Checked[idx]);
  }
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::UpdateComboBoxHistory (TComboBox *cb)
{
  int idx,i;

  if (cb->Text != "") {

    i = cb->Items->Count -1;

    while (i >= 0) {
      if (cb->Items->Strings[i] == cb->Text) {
        break;
      } else i--;
    }

    if (i < 0) {

      cb->Items->Insert (0, cb->Text);

    } else {

      if (i > 0) { cb->Items->Move (i, 0); }
    }

    cb->Text = cb->Items->Strings[0];
  }
}
//---------------------------------------------------------------------------
int __fastcall TBenchmarkXForm::ReadIntValue(TComboBox *cb, int min,int max)
{
  int Result;
  
  if (TryStrToInt (cb->Text, Result)) {
    if ((Result >= min) && (Result <= max)) {
      return Result;
    }
  } else {
    if (cb->CanFocus() ) { cb->SetFocus(); }
  }

  return min;
//  throw Exception("Expected value in range [%d..%d]", [min, max]);
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::btnRunClick(TObject *Sender)
{

//  Mtxvec::Controller->SetVecCacheSize(32,2048);
//  Mtxvec::Controller->SetMtxCacheSize(32,2048);

  TCursor cursor;
  int i;
  TBenchmarkResults *results;

  UpdateComboBoxHistory (cbVectorLength);
  UpdateComboBoxHistory (cbIterationCount);

  cursor = Screen->Cursor;
  Screen->Cursor = crHourGlass;
  try  {
    fBenchmarkFramework->VectorLength = ReadIntValue (cbVectorLength, 0, 1000000);
    fBenchmarkFramework->IterationCount = ReadIntValue (cbIterationCount, 1, 1000000);

    results = fBenchmarkResults;

    results->Clear();
    for (i = 0; i < clbFuncs->Count; i++) {
      if (clbFuncs->Checked[i]) {

        fBenchmarkFramework->Run (clbFuncs->Items->Strings[i], results);
      }
    }

    results->Title = Format( "CB6:  Length: %d   Count: %d ",
      OPENARRAY(TVarRec,(fBenchmarkFramework->VectorLength, fBenchmarkFramework->IterationCount)));

    ShowResults(results);
  }
  __finally {
    Screen->Cursor= cursor;
  }
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::btnOpenClick(TObject *Sender)
{
       LoadFile("");   
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::LoadFile(const AnsiString file_name)
{
   if (FileExists (file_name)) {

      fBenchmarkResults->LoadFromFile(file_name);
      UpdateChart();

  } else {

    LoadFromFileItemClick(NULL);
  }
}
//-----
void __fastcall TBenchmarkXForm::RecalcIterationCount()
{
  TCursor cursor;

  cursor = Screen->Cursor;
  Screen->Cursor = crHourGlass;
  try {
  
    cbIterationCount->Text = IntToStr(fBenchmarkFramework->RecalcIterationCount());
  }
  __finally {

    Screen->Cursor = cursor;
  }
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::FormActivate(TObject *Sender)
{
  if (fFirstActivation) {

    fFirstActivation = false;
    Application->ProcessMessages();
    RecalcIterationCount();
  }
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::LoadFromFileItemClick(TObject *Sender)
{
   if (OpenDialog1->Execute()) {
   
    fBenchmarkResults->LoadFromFile(OpenDialog1->FileName);
    UpdateChart();
  }
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::SaveToFileItemClick(TObject *Sender)
{
  if (SaveDialog1->Execute())  fBenchmarkResults->SaveToFile(SaveDialog1->FileName);
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::miSelectAllFuncsClick(TObject *Sender)
{
  SelectAllFuncs (true);
}
//---------------------------------------------------------------------------

void __fastcall TBenchmarkXForm::miUnselectAllFuncsClick(TObject *Sender)
{
  SelectAllFuncs (false);
}
//---------------------------------------------------------------------------
void __fastcall TBenchmarkXForm::UpdateChart()
{
  int i;

  Series1->Clear();
  Series2->Clear();
  Series3->Clear();
  Series4->Clear();

  chartResults->Height = 100 + (60 * fBenchmarkResults->Count);
  Caption = fBenchmarkResults->Title;

  for (i = fBenchmarkResults->Count-1; i > -1; i--) {

    Series1->AddBar(fBenchmarkResults->Items[i]->SmplVecTicks, fBenchmarkResults->Items[i]->FuncName, clDefault);
    Series2->AddBar(fBenchmarkResults->Items[i]->CplxVecTicks, fBenchmarkResults->Items[i]->FuncName, clDefault);
    Series3->AddBar(fBenchmarkResults->Items[i]->SmplFuncTicks, fBenchmarkResults->Items[i]->FuncName, clDefault);
    Series4->AddBar(fBenchmarkResults->Items[i]->CplxFuncTicks, fBenchmarkResults->Items[i]->FuncName, clDefault);
  }
}

void __fastcall TBenchmarkXForm::cbIterationCountDblClick(TObject *Sender)
{
  fBenchmarkFramework->VectorLength = ReadIntValue (cbVectorLength, 0, 1000000);
  RecalcIterationCount();
}
//---------------------------------------------------------------------------

File: MtxVec.Delphi/BenchmarkX.pas
unit BenchmarkX;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, Basic2, StdCtrls, ComCtrls, ExtCtrls, Menus, CheckLst, TeEngine,
  Series, TeeProcs, Chart, BenchmarkResults, BenchmarkFramework;

type
  TBenchmarkXForm = class(TBasicForm2)                   
    PopupMenu: TPopupMenu;                                 
    miSelectAllFuncs: TMenuItem;
    miUnselectAllFuncs: TMenuItem;      
    btnRun: TButton;                                            
    btnOpen: TButton;
    Label1: TLabel;
    cbVectorLength: TComboBox;
    cbIterationCount: TComboBox;
    Label2: TLabel;
    OpenDialog1: TOpenDialog;
    SaveDialog1: TSaveDialog;
    chartResults: TChart;
    Series4: THorizBarSeries;
    Series3: THorizBarSeries;
    Series2: THorizBarSeries;
    Series1: THorizBarSeries;
    Panel4: TPanel;
    clbFuncs: TCheckListBox;
    Label3: TLabel;
    ChartMenu: TPopupMenu;
    LoadFromFileItem: TMenuItem;
    SaveToFileItem: TMenuItem;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure clbFuncsDblClick(Sender: TObject);
    procedure btnRunClick(Sender: TObject);
    procedure btnOpenClick(Sender: TObject);
    procedure cbIterationCountDblClick(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure miSelectAllFuncsClick(Sender: TObject);
    procedure miUnselectAllFuncsClick(Sender: TObject);
    procedure LoadFromFileItemClick(Sender: TObject);
    procedure SaveToFileItemClick(Sender: TObject);
  private
    procedure LoadFuncListbox;
    function ReadIntValue(cb: TCombobox; min, max: integer): integer;
    procedure RecalcIterationCount;
    procedure UpdateComboboxHistory(cb: TCombobox);
    procedure miLoadFromFileClick(Sender: TObject);
//    procedure miSaveToFileClick(Sender: TObject);
    procedure ShowResults(results: TBenchmarkResults);
    procedure UpdateChart;
    procedure LoadFile(const file_name: string);
    procedure SelectAllFuncs(select: boolean);
    { Private declarations }
  public
    { Public declarations }
    fBenchmarkFramework: TBenchmarkFramework;
    fBenchmarkResults: TBenchmarkResults;
    fFirstActivation: boolean;
  end;

var
  BenchmarkXForm: TBenchmarkXForm;

implementation

uses MtxVecInt, MtxVec;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TBenchmarkXForm.FormCreate(Sender: TObject);
begin
  inherited;
  fFirstActivation:= TRUE;
  fBenchmarkFramework:= TBenchmarkFramework.Create;
  fBenchmarkResults:= TBenchmarkResults.Create;
  LoadFuncListbox;

  With RichEdit1.Lines do
  begin
    Clear;
    Add('These benchmarks compare vectorized and not vectorized ' +
        'versions of real and complex versions of basic math functions. ' +
        'The vectorized versions are labeled with MtxVec and the non-vectorized '+
        'versions are labeled with Math387.');
    Add('');
    Add('The non-vectorized real and complex versions ' +
        'are written in assembler for maximum performance. '+
        'Despite this, the vectorized versions are substantially faster. '+
        'Before running the tests double click on the iterations combobox. ' +
        'That will compute the number of iterations that have to be performed ' +
        'to get sufficiently long timings. The resolution of the timings is ' +
        'about 15ms.');
  end;
end;

procedure TBenchmarkXForm.FormDestroy(Sender: TObject);
begin
  fBenchmarkFramework.Free;
  fBenchmarkResults.Free;
  inherited;
end;

procedure TBenchmarkXForm.ShowResults (results: TBenchmarkResults);
begin
//  fBenchmarkResults.Assign(results);
  UpdateChart;
//  Show;
end;

procedure TBenchmarkXForm.LoadFuncListbox;
var
  i: integer;
begin
  clbFuncs.Items.BeginUpdate;
  try
    clbFuncs.Items.Clear;
    for i:= 0 to fBenchmarkFramework.FuncCount-1 do begin
      clbFuncs.Items.Add (fBenchmarkFramework.FuncName[i]);
      clbFuncs.Checked[i]:= TRUE;
    end;
  finally
    clbFuncs.Items.EndUpdate;
  end;
end;

procedure TBenchmarkXForm.SelectAllFuncs(select: boolean);
var
  i: integer;
begin
  for i:= 0 to clbFuncs.Items.Count-1 do
    clbFuncs.Checked[i]:= select;
end;

procedure TBenchmarkXForm.clbFuncsDblClick(Sender: TObject);
var
  idx: integer;
begin
  idx:= clbFuncs.ItemIndex;
  if idx <> -1 then
    clbFuncs.Checked[idx]:= not clbFuncs.Checked[idx];
end;

procedure TBenchmarkXForm.UpdateComboboxHistory (cb: TCombobox);
var
  i : integer;
begin
  if cb.Text <> '' then begin
    i:= cb.Items.Count-1;
    while (i >= 0) do
      if (cb.Items[i] = cb.Text) then
        break
      else
        Dec(i);

    if (i < 0) then
      cb.Items.Insert (0, cb.Text)
    else
      if (i > 0) then
        cb.Items.Move (i, 0);

    cb.Text:= cb.Items[0];
  end;
end;

function TBenchmarkXForm.ReadIntValue (cb: TCombobox; min, max: integer): integer;
begin
  if TryStrToInt (cb.Text, Result) then begin
    if (Result >= min) and (Result <= max) then
      Exit;
  end else
    if cb.CanFocus then
      cb.SetFocus;

  raise Exception.CreateFmt ('Expected value in range [%d..%d]', [min, max]);
end;

procedure TBenchmarkXForm.btnRunClick(Sender: TObject);
var
  cursor: TCursor;
  i: integer;
  results: TBenchmarkResults;
begin
  UpdateComboboxHistory (cbVectorLength);
  UpdateComboboxHistory (cbIterationCount);
                   
  cursor:= Screen.Cursor;
  Screen.Cursor:= crHourGlass;  
  try
    fBenchmarkFramework.VectorLength:= ReadIntValue (cbVectorLength, 0, 1000000);
    fBenchmarkFramework.IterationCount:= ReadIntValue (cbIterationCount, 1, 1000000);

    results := fBenchmarkResults;

    results.Clear;
    for i:= 0 to clbFuncs.Count-1 do
      if clbFuncs.Checked[i] then
        fBenchmarkFramework.Run (clbFuncs.Items[i], results);

    results.Title:= Format ('D7:  Length: %d   Count: %d ',
      [fBenchmarkFramework.VectorLength, fBenchmarkFramework.IterationCount]);

    ShowResults(results);

  finally
    Screen.Cursor:= cursor;
  end;
end;

procedure TBenchmarkXForm.btnOpenClick(Sender: TObject);
begin
    LoadFile('');
end;

procedure TBenchmarkXForm.LoadFile(const file_name: string);
begin
  if FileExists (file_name) then begin
    fBenchmarkResults.LoadFromFile(file_name);
    UpdateChart;
  end else
    miLoadFromFileClick(nil);
end;

procedure TBenchmarkXForm.cbIterationCountDblClick(Sender: TObject);
begin
  fBenchmarkFramework.VectorLength := ReadIntValue (cbVectorLength, 0, 1000000);
  RecalcIterationCount;
end;

procedure  TBenchmarkXForm.RecalcIterationCount;
var
  cursor: TCursor;
begin
  cursor:= Screen.Cursor;
  Screen.Cursor:= crHourGlass;
  try
    cbIterationCount.Text:= IntToStr(fBenchmarkFramework.RecalcIterationCount);
  finally
    Screen.Cursor:= cursor;
  end;
end;

procedure TBenchmarkXForm.FormActivate(Sender: TObject);
begin
  if fFirstActivation then begin
    fFirstActivation:= FALSE;
    Application.ProcessMessages;
    RecalcIterationCount;
  end;
end;

procedure TBenchmarkXForm.miLoadFromFileClick(Sender: TObject);
begin
  if OpenDialog1.Execute then begin
    fBenchmarkResults.LoadFromFile(OpenDialog1.FileName);
    UpdateChart;
  end;
end;

{procedure TBenchmarkXForm.miSaveToFileClick(Sender: TObject);
begin
  if SaveDialog1.Execute then
    fBenchmarkResults.SaveToFile(SaveDialog1.FileName);
end;}

procedure TBenchmarkXForm.miSelectAllFuncsClick(Sender: TObject);
begin
  SelectAllFuncs (TRUE);
end;

procedure TBenchmarkXForm.miUnselectAllFuncsClick(Sender: TObject);
begin
  SelectAllFuncs (false);
end;

procedure TBenchmarkXForm.LoadFromFileItemClick(Sender: TObject);
begin
  if OpenDialog1.Execute then begin
    fBenchmarkResults.LoadFromFile(OpenDialog1.FileName);
    UpdateChart;
  end;
end;

procedure TBenchmarkXForm.SaveToFileItemClick(Sender: TObject);
begin
  if SaveDialog1.Execute then
    fBenchmarkResults.SaveToFile(SaveDialog1.FileName);
end;

procedure TBenchmarkXForm.UpdateChart;
var
  i: integer;
begin
  Series1.Clear;
  Series2.Clear;
  Series3.Clear;
  Series4.Clear;

  chartResults.Height:= 100 + (60 * fBenchmarkResults.Count);
  Caption:= fBenchmarkResults.Title;

  for i:= fBenchmarkResults.Count-1 downto 0 do begin
    Series1.AddBar(fBenchmarkResults[i].SmplVecTicks, fBenchmarkResults[i].FuncName, clDefault);
    Series2.AddBar(fBenchmarkResults[i].CplxVecTicks, fBenchmarkResults[i].FuncName, clDefault);
    Series3.AddBar(fBenchmarkResults[i].SmplFuncTicks, fBenchmarkResults[i].FuncName, clDefault);
    Series4.AddBar(fBenchmarkResults[i].CplxFuncTicks, fBenchmarkResults[i].FuncName, clDefault);
  end;
end;

initialization
   RegisterClass(TBenchmarkXForm);

end.
